Vulnerability producer crashes during parsing of statements files, then restarts everything

Repro:
- Clone `git@github.com:fasten-project/fasten-docker-deployment.git`
- Checkout branch `add-vulnerability-producer-and-consumer`
- Put Github token in `fasten-vulnerability-producer.env`
- `docker-compose --profile security up -d`
- Observe log of Docker container: `fasten-docker-deployment_fasten-vulnerability-producer_1`

```
[2021-04-01 15:07:10,612] [INFO ] [main] [e.f.v.u.p.ExtraParser] - Parsing statement file for CVE-2013-4286
Cannot create property=notes for JavaBean=eu.fasten.vulnerabilityproducer.utils.mappers.YAMLHandler$SAPVulnMapper@79c5460e
in 'string', line 1, column 1:
vulnerability_id: CVE-2013-4286
^
Cannot create property=links for JavaBean=eu.fasten.vulnerabilityproducer.utils.mappers.YAMLHandler$NoteTextMapper@2e73d5eb
in 'string', line 3, column 3:
- links: []
^
Unable to find property 'links' on class: eu.fasten.vulnerabilityproducer.utils.mappers.YAMLHandler$NoteTextMapper
in 'string', line 3, column 10:
- links: []
^
in 'string', line 3, column 1:
- links: []
^
at org.yaml.snakeyaml.constructor.Constructor$ConstructMapping.constructJavaBean2ndStep(Constructor.java:292)
at org.yaml.snakeyaml.constructor.Constructor$ConstructMapping.construct(Constructor.java:171)
at org.yaml.snakeyaml.constructor.Constructor$ConstructYamlObject.construct(Constructor.java:331)
at org.yaml.snakeyaml.constructor.BaseConstructor.constructObjectNoCheck(BaseConstructor.java:230)
at org.yaml.snakeyaml.constructor.BaseConstructor.constructObject(BaseConstructor.java:219)
at org.yaml.snakeyaml.constructor.BaseConstructor.constructDocument(BaseConstructor.java:173)
at org.yaml.snakeyaml.constructor.BaseConstructor.getSingleData(BaseConstructor.java:157)
at org.yaml.snakeyaml.Yaml.loadFromReader(Yaml.java:472)
at org.yaml.snakeyaml.Yaml.load(Yaml.java:398)
at eu.fasten.vulnerabilityproducer.utils.parsers.ExtraParser.injectSAPProjectKb(ExtraParser.java:323)
at eu.fasten.vulnerabilityproducer.utils.parsers.ExtraParser.getVulnerabilities(ExtraParser.java:483)
at eu.fasten.vulnerabilityproducer.utils.parsers.ParserManager.getVulnerabilitiesFromParsers(ParserManager.java:92)
at eu.fasten.vulnerabilityproducer.VulnerabilityProducer.start(VulnerabilityProducer.java:105)
at eu.fasten.vulnerabilityproducer.Main.run(Main.java:118)
at picocli.CommandLine.executeUserObject(CommandLine.java:1729)
at picocli.CommandLine.access$900(CommandLine.java:145)
at picocli.CommandLine$RunLast.handle(CommandLine.java:2101)
at picocli.CommandLine$RunLast.handle(CommandLine.java:2068)
at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:1935)
at picocli.CommandLine.execute(CommandLine.java:1864)
at eu.fasten.vulnerabilityproducer.Main.main(Main.java:72)
Caused by: Cannot create property=links for JavaBean=eu.fasten.vulnerabilityproducer.utils.mappers.YAMLHandler$NoteTextMapper@2e73d5eb
in 'string', line 3, column 3:
- links: []
^
Unable to find property 'links' on class: eu.fasten.vulnerabilityproducer.utils.mappers.YAMLHandler$NoteTextMapper
in 'string', line 3, column 10:
- links: []
^
at org.yaml.snakeyaml.constructor.Constructor$ConstructMapping.constructJavaBean2ndStep(Constructor.java:292)
at org.yaml.snakeyaml.constructor.Constructor$ConstructMapping.construct(Constructor.java:171)
at org.yaml.snakeyaml.constructor.BaseConstructor.constructObjectNoCheck(BaseConstructor.java:230)
at org.yaml.snakeyaml.constructor.BaseConstructor.constructObject(BaseConstructor.java:219)
at org.yaml.snakeyaml.constructor.BaseConstructor.constructSequenceStep2(BaseConstructor.java:390)
at org.yaml.snakeyaml.constructor.BaseConstructor.constructSequence(BaseConstructor.java:374)
at org.yaml.snakeyaml.constructor.Constructor$ConstructSequence.construct(Constructor.java:538)
at org.yaml.snakeyaml.constructor.BaseConstructor.constructObjectNoCheck(BaseConstructor.java:230)
at org.yaml.snakeyaml.constructor.BaseConstructor.constructObject(BaseConstructor.java:219)
at org.yaml.snakeyaml.constructor.Constructor$ConstructMapping.newInstance(Constructor.java:305)
at org.yaml.snakeyaml.constructor.Constructor$ConstructMapping.constructJavaBean2ndStep(Constructor.java:268)
... 20 more

Caused by: org.yaml.snakeyaml.error.YAMLException: Unable to find property 'links' on class: eu.fasten.vulnerabilityproducer.utils.mappers.YAMLHandler$NoteTextMapper
at org.yaml.snakeyaml.introspector.PropertyUtils.getProperty(PropertyUtils.java:159)
at org.yaml.snakeyaml.introspector.PropertyUtils.getProperty(PropertyUtils.java:148)
at org.yaml.snakeyaml.constructor.Constructor$ConstructMapping.getProperty(Constructor.java:309)
at org.yaml.snakeyaml.constructor.Constructor$ConstructMapping.constructJavaBean2ndStep(Constructor.java:230)
... 30 more
[2021-04-01 15:07:10,630] [DEBUG] [Thread-0] [o.d.n.Nitrite] - Store compaction is successful.
[2021-04-01 15:07:10,632] [INFO ] [Thread-0] [o.d.n.Nitrite] - Nitrite database has been closed successfully.
```

--- a/src/main/java/eu/fasten/vulnerabilityproducer/utils/mappers/YAMLHandler.java
+++ b/src/main/java/eu/fasten/vulnerabilityproducer/utils/mappers/YAMLHandler.java
@@ -481,7 +481,7 @@ public class YAMLHandler {
         }
 
         public String vulnerability_id;
-        public List<NoteTextMapper> notes;
+        public List<NoteMapper> notes;
         public List<FixMapper> fixes;
         public List<ArtifactMapper> artifacts;
 
@@ -493,11 +493,11 @@ public class YAMLHandler {
             this.vulnerability_id = vulnerability_id;
         }
 
-        public List<NoteTextMapper> getNotes() {
+        public List<NoteMapper> getNotes() {
             return notes;
         }
 
-        public void setNotes(List<NoteTextMapper> notes) {
+        public void setNotes(List<NoteMapper> notes) {
             this.notes = notes;
         }
 

--- a/src/main/java/eu/fasten/vulnerabilityproducer/utils/parsers/ExtraParser.java
+++ b/src/main/java/eu/fasten/vulnerabilityproducer/utils/parsers/ExtraParser.java
@@ -94,24 +94,25 @@ public class ExtraParser implements VulnerabilityParser {
 
             while (keys.hasNext()) {
                 var key = keys.next();
-                var listVulns = (JSONArray) jsonObject.get(key);
-                for (Object listVuln : listVulns) {
-                    var vulnObject = (JSONObject) listVuln;
-                    var cveIds = vulnObject.get("cve");
-                    if (!cveIds.toString().equals("null")) {
-                        // There are multiple CVE-IDs
-                        if (cveIds.toString().contains(",")) {
-                            var cves = cveIds.toString().split(",\\s?");
-                            Arrays.stream(cves).forEach(cve -> insertVulnerability(injectInfoHelper(key, cve, vulnObject)));
+                var listVulns = jsonObject.optJSONArray(key);
+                if(Objects.nonNull(listVulns)) {
+                    for (Object listVuln : listVulns) {
+                        var vulnObject = (JSONObject) listVuln;
+                        var cveIds = vulnObject.get("cve");
+                        if (!cveIds.toString().equals("null")) {
+                            // There are multiple CVE-IDs
+                            if (cveIds.toString().contains(",")) {
+                                var cves = cveIds.toString().split(",\\s?");
+                                Arrays.stream(cves).forEach(cve -> insertVulnerability(injectInfoHelper(key, cve, vulnObject)));
+                            } else {
+                                // There is only one CVE-ID
+                                insertVulnerability(injectInfoHelper(key, cveIds.toString(), vulnObject));
+                            }
                         } else {
-                            // There is only one CVE-ID
-                            var cveId = cveIds;
-                            insertVulnerability(injectInfoHelper(key, cveId.toString(), vulnObject));
+                            // There is only PyUp-id
+                            var pyupId = (String) vulnObject.get("id");
+                            insertVulnerability(injectInfoHelper(key, pyupId, vulnObject));
                         }
-                    } else {
-                        // There is only PyUp-id
-                        var pyupId = (String) vulnObject.get("id");
-                        insertVulnerability(injectInfoHelper(key, pyupId, vulnObject));
                     }
                 }
             }

--- a/src/main/java/eu/fasten/vulnerabilityproducer/utils/parsers/NVDParser.java
+++ b/src/main/java/eu/fasten/vulnerabilityproducer/utils/parsers/NVDParser.java
@@ -41,6 +41,7 @@ import java.io.*;
 import java.net.MalformedURLException;
 import java.net.URL;
 import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
 import java.time.LocalDate;
 import java.util.*;
 import java.util.concurrent.TimeUnit;
@@ -104,15 +105,15 @@ public class NVDParser implements VulnerabilityParser {
      * @throws DownloadFailedException   - download fails
      * @throws ResourceNotFoundException - resource cannot be found
      */
-    public List<File> downloadCVEs() throws MalformedURLException,
+    public List<File> downloadCVEs() throws IOException,
             TooManyRequestsException,
-            DownloadFailedException,
             ResourceNotFoundException {
         int currentYear = LocalDate.now().getYear();
         List<File> paths = new ArrayList<>();
         for (int i = 2002; i <= currentYear; i++) {
             URL cve_url = new URL("https://nvd.nist.gov/feeds/json/cve/1.1/nvdcve-1.1-" + i + ".json.gz");
             File out_path = new File(mnt + "/nvd/nvdcve-1.1-" + i + ".json.gz");
+            Files.createDirectories(out_path.toPath());
             logger.info("Downloading NVD file for the year " + i);
             downloader.fetchFile(cve_url, out_path);
             paths.add(out_path);
@@ -129,12 +130,12 @@ public class NVDParser implements VulnerabilityParser {
      * @throws DownloadFailedException   - download fails
      * @throws ResourceNotFoundException - resource cannot be found
      */
-    public File downloadUpdates() throws MalformedURLException,
+    public File downloadUpdates() throws IOException,
             TooManyRequestsException,
-            DownloadFailedException,
             ResourceNotFoundException {
         var updates_url = new URL("https://nvd.nist.gov/feeds/json/cve/1.1/nvdcve-1.1-modified.json.gz");
         var out_path = new File(mnt + "/nvd/nvdcve-1.1-modified.json.gz");
+        Files.createDirectories(out_path.toPath());
         downloader.fetchFile(updates_url, out_path);
         return out_path;
     }
@@ -225,13 +226,7 @@ public class NVDParser implements VulnerabilityParser {
                 this.downloader = new Downloader(new Settings(new Properties()));
             }
             paths = downloadCVEs();
-        } catch (MalformedURLException e) {
-            e.printStackTrace();
-        } catch (TooManyRequestsException e) {
-            e.printStackTrace();
-        } catch (DownloadFailedException e) {
-            e.printStackTrace();
-        } catch (ResourceNotFoundException e) {
+        } catch (TooManyRequestsException | ResourceNotFoundException | IOException e) {
             e.printStackTrace();
         }
         List<DefCveItem> rawCves = new ArrayList<>();
@@ -264,13 +259,7 @@ public class NVDParser implements VulnerabilityParser {
                 this.downloader = new Downloader(new Settings(new Properties()));
             }
             pathToUpdates = downloadUpdates();
-        } catch (MalformedURLException e) {
-            e.printStackTrace();
-        } catch (TooManyRequestsException e) {
-            e.printStackTrace();
-        } catch (DownloadFailedException e) {
-            e.printStackTrace();
-        } catch (ResourceNotFoundException e) {
+        } catch (TooManyRequestsException | ResourceNotFoundException | IOException e) {
             e.printStackTrace();
         }
         List<DefCveItem> rawCves = new ArrayList<>();

--- a/src/main/java/eu/fasten/vulnerabilityproducer/utils/parsers/NVDParser.java
+++ b/src/main/java/eu/fasten/vulnerabilityproducer/utils/parsers/NVDParser.java
@@ -113,7 +113,7 @@ public class NVDParser implements VulnerabilityParser {
         for (int i = 2002; i <= currentYear; i++) {
             URL cve_url = new URL("https://nvd.nist.gov/feeds/json/cve/1.1/nvdcve-1.1-" + i + ".json.gz");
             File out_path = new File(mnt + "/nvd/nvdcve-1.1-" + i + ".json.gz");
-            Files.createDirectories(out_path.toPath());
+            Files.createDirectories(out_path.toPath().getParent());
             logger.info("Downloading NVD file for the year " + i);
             downloader.fetchFile(cve_url, out_path);
             paths.add(out_path);
@@ -135,7 +135,7 @@ public class NVDParser implements VulnerabilityParser {
             ResourceNotFoundException {
         var updates_url = new URL("https://nvd.nist.gov/feeds/json/cve/1.1/nvdcve-1.1-modified.json.gz");
         var out_path = new File(mnt + "/nvd/nvdcve-1.1-modified.json.gz");
-        Files.createDirectories(out_path.toPath());
+        Files.createDirectories(out_path.toPath().getParent());
         downloader.fetchFile(updates_url, out_path);
         return out_path;
     }

--- a/src/test/java/eu/fasten/vulnerabilityproducer/parsers/ExtraParserTest.java
+++ b/src/test/java/eu/fasten/vulnerabilityproducer/parsers/ExtraParserTest.java
@@ -213,7 +213,7 @@ public class ExtraParserTest {
     public void translateYAMLToVulnerability() {
         YAMLHandler.SAPVulnMapper yaml = new YAMLHandler.SAPVulnMapper();
         yaml.vulnerability_id = "CVE-TEST";
-        var note = new YAMLHandler.NoteTextMapper();
+        var note = new YAMLHandler.NoteMapper();
         note.text = "text";
         yaml.notes.add(note);
         var fixes = new YAMLHandler.FixMapper();

